src/main/kotlin/com/ampnet/identityservice/service/impl/VerificationServiceImpl.kt

```diff
@@ -0,0 +1,65 @@
+package com.ampnet.identityservice.service.impl
+
+import com.ampnet.identityservice.exception.ErrorCode
+import com.ampnet.identityservice.exception.InvalidRequestException
+import com.ampnet.identityservice.exception.ResourceNotFoundException
+import com.ampnet.identityservice.service.VerificationService
+import mu.KLogging
+import org.kethereum.crypto.signedMessageToKey
+import org.kethereum.crypto.toAddress
+import org.kethereum.model.SignatureData
+import org.springframework.stereotype.Service
+import java.math.BigInteger
+import java.security.SecureRandom
+import java.security.SignatureException
+import kotlin.jvm.Throws
+
+@Service
+@Suppress("MagicNumber")
+class VerificationServiceImpl : VerificationService {
+
+    companion object : KLogging()
+
+    private val userPayload = mutableMapOf<String, String>()
+
+    override fun generatePayload(address: String): String {
+        // TODO generates a negative number, find a function in kethereum
+        val nonce = SecureRandom().nextLong().toString()
+        userPayload[address] = nonce
+        return nonce
+    }
+
+    @Throws(ResourceNotFoundException::class, InvalidRequestException::class)
+    override fun verifyPayload(address: String, signedPayload: String): Boolean {
+        val payload = userPayload[address]
+            ?: throw ResourceNotFoundException(
+                ErrorCode.AUTH_PAYLOAD_MISSING, "There is no payload associated with address: $address."
+            )
+        userPayload.remove(address)
+        try {
+            val key = signedMessageToKey(payload.toByteArray(), getSignatureData(signedPayload))
+            return address == key.toAddress().toString()
+        } catch (ex: SignatureException) {
+            throw InvalidRequestException(
+                ErrorCode.AUTH_SIGNED_PAYLOAD_INVALID, "Public key cannot be recovered from the signature", ex
+            )
+        }
+    }
+
+    private fun getSignatureData(signedPayload: String): SignatureData {
+        if (signedPayload.length != 132)
+            throw InvalidRequestException(
+                ErrorCode.AUTH_SIGNED_PAYLOAD_INVALID, "Signature: $signedPayload is of wrong length."
+            )
+        val r = signedPayload.substring(2, 66)
```

Comment:
Are there any better names instead of `r`, `s` and `v`?

Reply:
Not really, these are just point calculated from the Elliptic Curve(ECDSA)

Reply:
Ok, in that case it would be nice to have a comment before them which says something like `Elliptic curve points` or similar.

Reply:
Sure, I'll put some comments.